### PR TITLE
Remove python 3.8 support and add python 3.13 support

### LINKED ISSUE #1988: [Feature]: Python 3.8
### What would you like to see added to PyNWB?

Python 3.8 is deprecated. Update the pyproject.toml

### Is your feature request related to a problem?

_No response_

### What solution would you like?

Above

### Do you have any interest in helping implement the feature?

Yes.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/.github/CODE_OF_CONDUCT.rst)
- [X] Have you checked the [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured this change was not already [requested](https://github.com/NeurodataWithoutBorders/pynwb/issues)?

### PR DESCRIPTION
**Pull request by stephprince**
## Motivation

Fix https://github.com/NeurodataWithoutBorders/pynwb/issues/1988. Python 3.8 has been deprecated - we will remove support for 3.8 and add support for python 3.13.

TODO
- [x] update GitHub branch protections to switch to new workflow minimums before merging 

## Checklist

- [x] Did you update CHANGELOG.md with your changes?
- [X] Have you checked our [Contributing](https://github.com/NeurodataWithoutBorders/pynwb/blob/dev/docs/CONTRIBUTING.rst) document?
- [X] Have you ensured the PR clearly describes the problem and the solution?
- [X] Is your contribution compliant with our coding style? This can be checked running `ruff check . && codespell` from the source directory.
- [X] Have you checked to ensure that there aren't other open [Pull Requests](https://github.com/NeurodataWithoutBorders/pynwb/pulls) for the same change?
- [X] Have you included the relevant issue number using "Fix #XXX" notation where XXX is the issue number? By including "Fix #XXX" you allow GitHub to close issue #XXX when the PR is merged.

### CODE REVIEW

**Comment by stephprince** [comment]
I believe the python-3.13 gallery tests are failing because `requirements-doc.txt` requires dandi, which requires nwbinspector, which currently has `h5py<3.12.0` pinned. h5py 3.12 is the earliest release with pre-built packages for python 3.13. Until h5py is unpinned in the inspector, using python 3.13 with the current requirements will try to build the h5py wheel. 

https://github.com/NeurodataWithoutBorders/nwbinspector/pull/536 is in progress to remove the h5py restriction in the inspector. We can finish that PR and make another nwbinspector release, or use 3.12 for the gallery and dandi tests, or fix the h5py build.

The cause of the excessive warnings in the gallery-3.9-minimum tests is less clear to me…. it seems to be coming from one of the dependencies for the `linkml-runtime`, `schema sheets`, or `oaklib` packages, which were not previously being installed with python 3.8. Unpinning the packages does not solve the warnings when using python 3.9.

**Comment by rly** [comment]
Let's make another nwbinspector release. 

Interesting. It looks like we have not actually been testing against minimum requirements. See https://github.com/NeurodataWithoutBorders/pynwb/actions/runs/12024183637/job/33519287151

It installs the minimum requirements first and then upgrades them.
```
py38-minimum: install_deps> python -m pip install -U -r requirements-dev.txt -r requirements-min.txt
.pkg: install_requires> python -I -m pip install hatch-vcs hatchling
.pkg: _optional_hooks> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: get_requires_for_build_sdist> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True hatchling.build
.pkg: freeze> python -m pip freeze --all
.pkg: hatch-vcs==0.4.0,hatchling==1.26.3,packaging==24.2,pathspec==0.12.1,pip==24.3.1,pluggy==1.5.0,setuptools==75.3.0,setuptools-scm==8.1.0,tomli==2.1.0,trove-classifiers==2024.10.21.16,typing_extensions==4.12.2,wheel==0.45.1
.pkg: build_sdist> python /opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/pyproject_api/_backend.py True hatchling.build
py38-minimum: install_package_deps> python -m pip install -U 'h5py>=2.10' 'hdmf>=3.14.5' 'numpy>=1.18' 'pandas>=1.1.5' 'python-dateutil>=2.7.3'
```

There should not be a `-U` on this line https://github.com/NeurodataWithoutBorders/pynwb/blob/a63387d9068a0c9e5ace1839a1fafba4d3d18e34/tox.ini#L16

You could also change `tox.ini` to this abstracted, easier to update, but potentially more confusing `tox.ini` format that we use in HDMF: https://github.com/hdmf-dev/hdmf/blob/dev/tox.ini

**Comment by stephprince** [comment]
Oh interesting, I think that `tox.ini` format you linked looks nice and easier to maintain. I will try updating to use that instead.

**Comment by stephprince** [comment]
>The cause of the excessive warnings in the gallery-3.9-minimum tests is less clear to me…. it seems to be coming from one of the dependencies for the linkml-runtime, schema sheets, or oaklib packages, which were not previously being installed with python 3.8. Unpinning the packages does not solve the warnings when using python 3.9.

I think there is something weird going on with the resolver between dependencies for the `linkml-runtime`, `schemasheets`, and `oaklib` packages and `dandi`. If you install `dandi` after installing everything else in `requirements-doc.txt` it installs fine with python 3.9. Everything also installs fine in python 3.10. I'm not sure if it's worth debugging more or, for the purposes of the minimum gallery tests, if it's ok to set the lower python bound to 3.10 for those extra packages as we had it set previously.

File: requirements-min.txt [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # minimum versions of package dependencies for installing PyNWB
-h5py==2.10  # support for selection of datasets with list of indices added in 2.10
+h5py==3.2.0
 hdmf==3.14.5
-numpy==1.18
-pandas==1.1.5
-python-dateutil==2.7.3
+numpy==1.23.0
```
Comment (stephprince) [inline requirements-min.txt]: numpy 1.23.0 introduced updates to [structured (void) dtype comparisons](https://numpy.org/doc/stable/release/1.23.0-notes.html#changes-to-structured-void-dtype-promotion-and-comparisons) used to compare the scalars with compound data type in `ElectrodeGroup.position`. 

If 1.23.0 is too high of a minimum, we could go back to 1.20.0 and handle the comparisons ourselves in [`_assert_array_equal`](https://github.com/hdmf-dev/hdmf/blob/d1d1b5a654a4d0887929770f0d48493ab99621da/src/hdmf/testing/testcase.py#L160-L167) in hdmf

Comment (rly) [reply requirements-min.txt]: I think this is OK. We can adjust if someone has a good reason to use numpy < 1.23.0

**Review by rly — APPROVED** [review]

**Comment by rly** [comment]
Looks good to me! Thanks.

**Pull request merged**